packages/next/src/server/lib/incremental-cache/index.ts

```diff
@@ -151,58 +179,79 @@ export class IncrementalCache {
   async fetchCacheKey(
     url: string,
-    init: RequestInit | Request = {}
-  ): Promise<string> {
+    init: RequestInit | Request | undefined,
+    options: { returnBody: true }
+  ): Promise<{
+    cacheKey: string
+    body: Blob | ReadableStream | FormData | string | undefined
+  }>
+  async fetchCacheKey(
+    url: string,
+    init?: RequestInit | Request
+  ): Promise<string>
+  async fetchCacheKey(
+    url: string,
+    init: RequestInit | Request = {},
+    options?: { returnBody: true }
+  ): Promise<
+    | string
+    | {
+        cacheKey: string
+        body: Blob | ReadableStream | FormData | string | undefined
+      }
+  > {
     // this should be bumped anytime a fix is made to cache entries
     // that should bust the cache
     const MAIN_KEY_PREFIX = 'v1'
 
     let cacheKey: string
     const bodyChunks: string[] = []
 
+    let body: Blob | ReadableStream | FormData | string | undefined
     if (init.body) {
+      // handle string body
+      if (typeof init.body === 'string') {
+        bodyChunks.push(init.body)
+        body = init.body
+      }
       // handle ReadableStream body
-      if (typeof (init.body as any).getReader === 'function') {
-        const readableBody = init.body as ReadableStream
-        const reader = readableBody.getReader()
-        let arrayBuffer = new Uint8Array()
-
-        function processValue({
-          done,
-          value,
-        }: {
-          done?: boolean
-          value?: ArrayBuffer | string
-        }): any {
-          if (done) {
-            return
-          }
-          if (value) {
-            try {
-              bodyChunks.push(typeof value === 'string' ? value : encode(value))
-              const curBuffer: Uint8Array =
-                typeof value === 'string'
-                  ? encodeText(value)
-                  : new Uint8Array(value)
-
-              const prevBuffer = arrayBuffer
-              arrayBuffer = new Uint8Array(
-                prevBuffer.byteLength + curBuffer.byteLength
-              )
-              arrayBuffer.set(prevBuffer)
-              arrayBuffer.set(curBuffer, prevBuffer.byteLength)
-            } catch (err) {
-              console.error(err)
-            }
-          }
-          reader.read().then(processValue)
-        }
-        await reader.read().then(processValue)
-        ;(init as any)._ogBody = arrayBuffer
-      } // handle FormData or URLSearchParams bodies
-      else if (typeof (init.body as any).keys === 'function') {
-        const formData = init.body as FormData
-        ;(init as any)._ogBody = init.body
+      else if (
+        'getReader' in init.body &&
+        typeof init.body.getReader === 'function'
+      ) {
+        // We need to tee() the stream to get two readers. One for collecting
+        // the body for the cache key and one for passing back for fetch.
+        const [ogBodyReader, cacheKeyReader] = init.body.tee()
```

Comment:
Rather than manually assembing an `ArrayBuffer` from the stream, we can just `.tee()` the reader to split the readers up and consume one now to collect the chunks and pass back the tee'd stream to be consumed.